### PR TITLE
Bug 1569274 - Update Dockerfile and circleci Updated Dockerfile and circleci configuration to build dependencies into a new bmo-slim image based on Perl 5.30

### PR DESCRIPTION
**Pull request by dklawren**
All tests are now passing.

## Bug Link

[Bug 1569274 - Update Dockerfile and circleci config to build dependencies dynamically instead of relying on bmo-slim images](https://bugzilla.mozilla.org/show_bug.cgi?id=1569274)

### CODE REVIEW

**Comment by dylanwh** [comment]
Probably the right way forward for this would be to:

- include cpanfile and cpanfile.snapshot in the main repo
- install cpm using cpanm
- use cpm install to install all the dependencies.

This will be pretty fast and it's what I advised the people in singapore to use.

**Comment by dklawren** [comment]
> 
> 
> Probably the right way forward for this would be to:
> 
>     * include cpanfile and cpanfile.snapshot in the main repo
> 
>     * install cpm using cpanm
> 
>     * use cpm install to install all the dependencies.
> 
> 
> This will be pretty fast and it's what I advised the people in singapore to use.

Hi Dylan. Thanks for the feedback :) Are you suggesting to remove the 'perl Makefile.PL' and 'make cpanfile' steps in the first build stage and just let cpm install from the committed cpanfile? What is the advantage of this over just dynamically generating the cpanfile with 'make cpanfile' each commit to master and subsequent CI build?

**Comment by dylanwh** [comment]
> > Probably the right way forward for this would be to:
> > ```
> > * include cpanfile and cpanfile.snapshot in the main repo
> > 
> > * install cpm using cpanm
> > 
> > * use cpm install to install all the dependencies.
> > ```
> > 
> > 
> > This will be pretty fast and it's what I advised the people in singapore to use.
> 
> Hi Dylan. Thanks for the feedback :) Are you suggesting to remove the 'perl Makefile.PL' and 'make cpanfile' steps in the first build stage and just let cpm install from the committed cpanfile? What is the advantage of this over just dynamically generating the cpanfile with 'make cpanfile' each commit to master and subsequent CI build?

You want cpanfile and especially cpanfile.snapshot to be versioned with the repo to prevent unintentional upgrades. Remember that the `snapshot` one records the exact version of every dependency, recursively.

One additional thing to realize is that the OS and perl version impact the content of cpanfile.lock. This is a problem for upstream, but shouldn't be for BMO. When you switch distro or perl version, you'll need re-generate cpanfile.snapshot from scratch.

I do still get email pings for this repo, but if you need a faster response I'm still in IRC -- also in #bugzilla on freenode. :)

File: Dockerfile [hunk 1]
```diff
@@ -10,27 +38,40 @@ ENV CIRCLE_SHA1=${CIRCLE_SHA1}
...
-COPY . .
 
-RUN mv /opt/bmo/local /app && \
-    chown -R app:app /app && \
+COPY --from=builder /app/local /app/local
+COPY --from=builder /app/PACKAGES /app/PACKAGES
+
+RUN apt-get update && apt-get upgrade -y && apt-get install -y curl git libcap2-bin xz-utils vim $(cat PACKAGES)
+
+RUN curl -L https://github.com/dylanwh/tocotrienol/releases/download/1.0.6/tct-centos6.tar.xz > /usr/local/bin/tct.tar.xz && \
```
[the remark this answers is not in the document]
Comment (dklawren) [reply Dockerfile]: @glob any objection to just bringing the tct binary into the core BMO repo under extensions/SecureMail/bin/tct ?

We could still keep the mozilla-bteam/tocotrienol fork around to generate new binaries as needed. Having the binary included would get rid of the need to keep it in a S3 bucket altogether.

[the remark this answers is not in the document]
Comment (globau) [reply Dockerfile]: works for me.

File: Dockerfile [hunk 2]
```diff
@@ -10,27 +38,40 @@ ENV CIRCLE_SHA1=${CIRCLE_SHA1}
...
+    tar -C /usr/local/bin -xvf /usr/local/bin/tct.tar.xz && \
+    rm /usr/local/bin/tct.tar.xz && \
+    chmod +x /usr/local/bin/tct && \
+    curl -L https://github.com/krallin/tini/releases/download/${TINI_VERSION}/tini > /usr/local/sbin/tini && \
```
[the remark this answers is not in the document]
Comment (dklawren) [reply Dockerfile]: tct will be brought into the core code. tini was removed as it was not being used anywhere.

**Pull request merged**